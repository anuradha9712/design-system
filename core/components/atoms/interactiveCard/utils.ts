export const removeItem = (targetItem: any, arr = []) => {
  return arr.filter((item: any) => item !== targetItem);
};
