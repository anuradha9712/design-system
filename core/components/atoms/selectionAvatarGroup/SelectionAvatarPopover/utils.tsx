import React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>,
  listRef?: any,
  withSearch?: boolean
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      navigateOptions('up', focusedOption, setFocusedOption, listRef, withSearch);
      break;
    case 'ArrowDown':
      event.preventDefault();
      navigateOptions('down', focusedOption, setFocusedOption, listRef, withSearch);
      break;
    case 'Enter':
      handleEnterKey(focusedOption);
      setHighlightLastItem?.(false);
      setHighlightFirstItem?.(false);
      break;
    default:
      break;
  }
};

const handleEnterKey = (focusedOption: Element | undefined) => {
  (focusedOption as HTMLElement)?.click();
};

const navigateSelectedOption = (
  index: number,
  withSearch?: boolean,
  listItems?: any,
  direction?: string,
  listRef?: any,
  setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>
) => {
  console.log('tttindexxx', index);
  // if (withSearch) {
  //   const searchInput = listRef.current.querySelector('[data-test="DesignSystem-SelectionAvatar--Input"]');
  //   console.log('ttt inside iff');
  //   if ((index === 0 && direction === 'up') || (index === listItems.length - 2 && direction === 'down')) {
  //     console.log('tttfocus search');
  //     searchInput.focus();
  //     setFocusedOption && setFocusedOption(searchInput);
  //     return;
  //   }
  // }
  const searchInput = listRef.current.querySelector('[data-test="DesignSystem-SelectionAvatar--Input"]');
  searchInput.focus();
  setFocusedOption && setFocusedOption(searchInput);
};

const navigateOptions = (
  direction: string,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>,
  listRef?: any,
  withSearch?: boolean
) => {
  const listItems = listRef.current.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let index = Array.from(listItems).findIndex((item) => {
    return item == focusedOption;
  });

  console.log('tttwithSearchwithSearch', withSearch, 'index', index, 'listItems[index]', listItems[index]);

  if (index === -1) {
    console.log('ttt11');
    index = direction === 'up' ? listItems.length - 1 : 0;
  } else if (
    (withSearch && index === 0 && direction === 'up') ||
    (withSearch && index === listItems.length - 1 && direction === 'down')
  ) {
    console.log('ttt22');

    navigateSelectedOption(index, withSearch, listItems, direction, listRef, setFocusedOption);
    // return;
  } else {
    console.log('ttt333');

    index = direction === 'up' ? (index - 1 + listItems.length) % listItems.length : (index + 1) % listItems.length;

    const targetOption = listItems[index];

    (targetOption as HTMLElement).focus();
    setFocusedOption && setFocusedOption(targetOption);
    targetOption.scrollIntoView({ block: 'center' });
  }
};

export const handleInputKeyDown = (
  event: React.KeyboardEvent,
  listRef: any,
  setFocusedOption: React.Dispatch<React.SetStateAction<Element | undefined>>
) => {
  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;

  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      targetOption = listItems[listItems.length - 1];
      break;
    case 'ArrowDown':
      event.preventDefault();
      targetOption = listItems[0];
      break;
    default:
      break;
  }

  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};
