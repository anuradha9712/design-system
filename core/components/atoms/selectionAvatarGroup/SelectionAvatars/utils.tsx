import React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>
) => {
  switch (event.key) {
    case 'Enter':
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'ArrowDown':
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'ArrowUp':
      setOpenPopover?.(true);
      setHighlightLastItem?.(true);
      break;
    default:
      break;
  }
};

export const focusListItem = (
  position: string,
  setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>,
  listRef?: any,
  withSearch?: boolean
) => {
  if (withSearch && position === 'down') {
    const searchInput = listRef.current?.querySelectorAll('[data-test="DesignSystem-SelectionAvatar--Input"]');
    console.log('aa22', searchInput);
    searchInput[0]?.focus();
    setFocusedOption && setFocusedOption(searchInput);
    return;
  }

  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  console.log('listReflistRef', listRef.current, 'listItems', listItems);
  let targetOption;

  if (position === 'down') {
    targetOption = listItems[0];
  } else {
    targetOption = listItems[listItems.length - 1];
  }
  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};
