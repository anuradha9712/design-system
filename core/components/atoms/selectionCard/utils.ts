export const removeItem = (targetItem: string, arr: string[]) => {
  return arr.filter((item: string) => item !== targetItem);
};

export const handleMultiSelect = (
  selectedOptions: string[],
  isItemSelected: boolean,
  setSelectedOptions: React.Dispatch<React.SetStateAction<string[]>>,
  id: string
) => {
  let resultList = [...selectedOptions];
  if (isItemSelected) {
    resultList = removeItem(id, selectedOptions);
  } else {
    resultList.push(id);
  }
  setSelectedOptions(resultList);
};

export const handleSingleSelect = (
  selectedOptions: string[],
  isItemSelected: boolean,
  setSelectedOptions: React.Dispatch<React.SetStateAction<string[]>>,
  id: string
) => {
  let resultList = [...selectedOptions];
  if (isItemSelected) {
    resultList = [];
  } else {
    resultList = [id];
  }

  setSelectedOptions(resultList);
};
