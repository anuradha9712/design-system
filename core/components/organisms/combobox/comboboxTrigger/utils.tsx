import React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  // setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>,
  setHighlightFirstItem?: any,
  setHighlightLastItem?: any
) => {
  console.log('handle key down e.key', event.key);

  switch (event.key) {
    case 'ArrowUp':
      setOpenPopover?.(true);
      setHighlightLastItem(true);
      // focusListItem('up', setFocusedOption);
      // requestAnimationFrame(() => focusListItem('up', setFocusedOption));
      break;
    case 'ArrowDown':
      setOpenPopover?.(true);
      setHighlightFirstItem(true);

      // requestAnimationFrame(() => focusListItem('down', setFocusedOption));
      break;
    case 'Escape':
      setOpenPopover?.(false);
      break;
    default:
      break;
  }
};

export const focusListItem = (
  position: string,
  setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>
) => {
  const listItems = document.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;
  // if (position === 'down') {
  //   targetOption = listItems[0];
  //   (listItems[0] as HTMLElement)?.focus();
  // } else {
  //   (listItems[listItems.length - 1] as HTMLElement)?.focus();
  // }

  if (position === 'down') {
    console.log('rrr down', listItems);
    targetOption = listItems[0];
  } else {
    targetOption = listItems[listItems.length - 1];
  }
  console.log('rrrtargetOption', targetOption);
  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};
