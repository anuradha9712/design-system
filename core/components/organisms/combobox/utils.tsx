import React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  inputTriggerRef?: any
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      navigateOptions('up', focusedOption, setFocusedOption);
      break;
    case 'ArrowDown':
      event.preventDefault();
      navigateOptions('down', focusedOption, setFocusedOption);
      break;
    case 'Enter':
      handleEnterKey(focusedOption);
      break;
    case 'Escape':
      setOpenPopover?.(false);
      console.log('inputTriggerRef.current', inputTriggerRef.current);
      inputTriggerRef.current.focus();
      break;
    default:
      break;
  }
};

const handleEnterKey = (focusedOption: Element | undefined) => {
  (focusedOption as HTMLElement)?.click();
};

const navigateOptions = (
  direction: string,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<Element | undefined>>
) => {
  const listItems = document.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let index = Array.from(listItems).findIndex((item) => {
    return item == focusedOption;
  });

  if (index === -1) {
    index = direction === 'up' ? listItems.length - 1 : 0;
  } else {
    index = direction === 'up' ? (index - 1 + listItems.length) % listItems.length : (index + 1) % listItems.length;
  }

  const targetOption = listItems[index];
  console.log('targetOptiontargetOption', targetOption);
  (targetOption as HTMLElement).focus();
  setFocusedOption && setFocusedOption(targetOption);
  targetOption.scrollIntoView({ block: 'center' });
};
