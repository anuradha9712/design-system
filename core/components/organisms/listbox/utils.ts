const focusOption = (targetIndex: number, totalLength: number, direction: string) => {
  let startIndex = targetIndex;
  const endIndex = direction === 'down' ? totalLength : -1;

  while (startIndex !== endIndex) {
    const selectedElement = document.querySelector(`[data-list-item-index="${startIndex}"]`) as HTMLElement;

    if (selectedElement && selectedElement.getAttribute('data-disabled') !== 'true') {
      selectedElement && selectedElement.focus();
      break;
    }

    if (direction === 'down') {
      startIndex++;
    } else {
      startIndex--;
    }
  }
};

export const onKeyDown = (event: React.KeyboardEvent, totalLength: number) => {
  const sourceElement = event.target as HTMLElement;
  const sourceIndex = sourceElement.getAttribute('data-list-item-index');
  const targetIndex = (sourceIndex && parseInt(sourceIndex, 10)) || 0;

  switch (event.key) {
    case 'ArrowDown':
      event.preventDefault();
      focusOption(targetIndex + 1, totalLength, 'down');
      break;
    case 'ArrowUp':
      event.preventDefault();
      focusOption(targetIndex - 1, totalLength, 'up');
      break;
    case 'Enter': {
      event.preventDefault();
      sourceElement.click();
      break;
    }
    default:
      break;
  }
};
