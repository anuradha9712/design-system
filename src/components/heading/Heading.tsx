import * as React from 'react';
import GenericText from '../_text';
import classNames from 'classnames';
import { BaseHtmlProps, BaseProps } from '@/utils/types';
import { HeadingVariant, HeadingSize } from '@/common.type';

export interface HeadingProps extends BaseProps, BaseHtmlProps<HTMLHeadingElement> {
  /**
   * Text to be rendered
   */
  children: React.ReactText;
  /**
   * Defines the type of heading
   */
  element: HeadingVariant;
  /**
   * size of `Heading`
   */
  size: HeadingSize;
  /**
   * Makes `Heading` appearance as subtle
   */
  subtle?: boolean;
  /**
   * Makes `Heading` appearance as inverted
   */
  inverted?: boolean;
  /**
   * Makes `Heading` appearance as disabled
   */
  disabled?: boolean;
}

export const Heading = (props: HeadingProps) => {
  const { size, children, subtle, inverted, disabled, className, element, ...rest } = props;

  const classes = classNames(
    {
      ['Mds-Heading']: true,
      [`Mds-Heading--${size}`]: size,
      [`Mds-Heading--subtle`]: subtle,
      [`Mds-Heading--disabled`]: disabled,
      [`Mds-Heading--white`]: inverted,
    },
    className
  );

  return (
    <GenericText data-test="DesignSystem-Heading" {...rest} className={classes} componentType={element}>
      {children}
    </GenericText>
  );
};

Heading.displayName = 'Heading';
Heading.defaultProps = {
  // appearance: 'default',
  element: 'h3',
  size: 'medium',
};

export default Heading;
