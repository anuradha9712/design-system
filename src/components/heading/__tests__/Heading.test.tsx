import * as React from 'react';
import { render } from '@testing-library/react';
import Heading, { HeadingProps as Props, HeadingSize } from '../Heading';
import { testHelper, filterUndefined, valueHelper, testMessageHelper } from '@/utils/testHelper';
import { HeadingAppearance } from '@/common.type';

const appearance: HeadingAppearance[] = ['default', 'subtle', 'disabled', 'white'];
const sizes: HeadingSize[] = ['s', 'm', 'l', 'xl', 'xxl'];
const sizeMap = {
  s: 'H5',
  m: 'H4',
  l: 'H3',
  xl: 'H2',
  xxl: 'H1',
};

describe('Heading component Snapshot for prop:appearance', () => {
  const mapper = {
    appearance: valueHelper(appearance, { required: true, iterate: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { baseElement } = render(<Heading {...attr}>Heading</Heading>);
      expect(baseElement).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('Heading component Snapshot for prop:size', () => {
  const mapper = {
    size: valueHelper(sizes, { required: true, iterate: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { baseElement } = render(<Heading {...attr}>Heading</Heading>);
      expect(baseElement).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('Heading component Unit Tests', () => {
  it('renders children', () => {
    const { getByTestId } = render(<Heading>{'Heading'}</Heading>);
    expect(getByTestId('DesignSystem-Heading')).toHaveTextContent('Heading');
  });

  it('overwrite Heading class', () => {
    const { getByTestId } = render(<Heading className="HeadingClass">{'Heading'}</Heading>);
    expect(getByTestId('DesignSystem-Heading')).toHaveClass('HeadingClass');
  });
});

describe('Heading component with prop:appearance', () => {
  appearance.forEach((appear) => {
    it(`should have the Heading--${appear} class when appearance=${appear} `, () => {
      const { getByTestId } = render(<Heading appearance={appear}>{'Heading'}</Heading>);
      expect(getByTestId('DesignSystem-Heading')).toHaveClass(`Mds-Heading--${appear}`);
    });
  });
});

describe('Heading component with prop:size', () => {
  sizes.forEach((size) => {
    it(`should have the Heading--${size} class when appearance=${size} `, () => {
      const { getByTestId } = render(<Heading size={size}>{'Heading'}</Heading>);
      expect(getByTestId('DesignSystem-Heading')).toHaveClass(`Mds-Heading--${size}`);
    });
  });
});

describe('Heading component attribute componentType', () => {
  sizes.forEach((size) => {
    it(`should have the Heading tagname`, () => {
      const { getByTestId } = render(<Heading size={size}>{'Heading'}</Heading>);
      expect(getByTestId('DesignSystem-Heading').tagName).toEqual(`${sizeMap[size]}`);
    });
  });
});
