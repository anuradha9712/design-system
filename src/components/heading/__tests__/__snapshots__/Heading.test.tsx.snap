// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading component Snapshot for prop:appearance 
  appearance: "default"
 1`] = `
<body>
  <div>
    <h4
      class="Mds-Heading Mds-Heading--m Mds-Heading--default"
      data-test="DesignSystem-Heading"
    >
      Heading
    </h4>
  </div>
</body>
`;

exports[`Heading component Snapshot for prop:appearance 
  appearance: "disabled"
 1`] = `
<body>
  <div>
    <h4
      class="Mds-Heading Mds-Heading--m Mds-Heading--disabled"
      data-test="DesignSystem-Heading"
    >
      Heading
    </h4>
  </div>
</body>
`;

exports[`Heading component Snapshot for prop:appearance 
  appearance: "subtle"
 1`] = `
<body>
  <div>
    <h4
      class="Mds-Heading Mds-Heading--m Mds-Heading--subtle"
      data-test="DesignSystem-Heading"
    >
      Heading
    </h4>
  </div>
</body>
`;

exports[`Heading component Snapshot for prop:appearance 
  appearance: "white"
 1`] = `
<body>
  <div>
    <h4
      class="Mds-Heading Mds-Heading--m Mds-Heading--white"
      data-test="DesignSystem-Heading"
    >
      Heading
    </h4>
  </div>
</body>
`;

exports[`Heading component Snapshot for prop:size 
  size: "l"
 1`] = `
<body>
  <div>
    <h3
      class="Mds-Heading Mds-Heading--l Mds-Heading--default"
      data-test="DesignSystem-Heading"
    >
      Heading
    </h3>
  </div>
</body>
`;

exports[`Heading component Snapshot for prop:size 
  size: "m"
 1`] = `
<body>
  <div>
    <h4
      class="Mds-Heading Mds-Heading--m Mds-Heading--default"
      data-test="DesignSystem-Heading"
    >
      Heading
    </h4>
  </div>
</body>
`;

exports[`Heading component Snapshot for prop:size 
  size: "s"
 1`] = `
<body>
  <div>
    <h5
      class="Mds-Heading Mds-Heading--s Mds-Heading--default"
      data-test="DesignSystem-Heading"
    >
      Heading
    </h5>
  </div>
</body>
`;

exports[`Heading component Snapshot for prop:size 
  size: "xl"
 1`] = `
<body>
  <div>
    <h2
      class="Mds-Heading Mds-Heading--xl Mds-Heading--default"
      data-test="DesignSystem-Heading"
    >
      Heading
    </h2>
  </div>
</body>
`;

exports[`Heading component Snapshot for prop:size 
  size: "xxl"
 1`] = `
<body>
  <div>
    <h1
      class="Mds-Heading Mds-Heading--xxl Mds-Heading--default"
      data-test="DesignSystem-Heading"
    >
      Heading
    </h1>
  </div>
</body>
`;
